From e1e1bfcaec73a27fbc6c6b31af145a68398f42c1 Mon Sep 17 00:00:00 2001
From: Christian Lamparter <chunkeey@googlemail.com>
Date: Fri, 29 Mar 2013 18:41:14 +0100
Subject: [PATCH 05/65] r92su: fix panic when adding mic/icv
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The network core does not guarantee that every frame has the
needed headroom and tailroom available. So, the driver has
to check whenever there's enough required free room for the
extra data we add.

Reported-by: José Vázquez <ppvazquezfer@gmail.com>
Signed-off-by: Christian Lamparter <chunkeey@googlemail.com>
---
 r92su/tx.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

--- a/drivers/net/wireless/r92su/tx.c
+++ b/drivers/net/wireless/r92su/tx.c
@@ -278,6 +278,21 @@ r92su_tx_prepare_tx_info_and_find_sta(st
 	struct r92su_tx_info *tx_info = r92su_get_tx_info(skb);
 	struct r92su_sta *sta;
 	struct ethhdr *hdr = (void *) skb->data;
+	int needed_tailroom;
+
+	/* The network core does not guarantee that every frame has the
+	 * needed headroom and tailroom available. So, the driver has
+	 * to check whenever there's enough required free room for the
+	 * extra data we add.
+	 */
+	needed_tailroom = r92su->wdev.netdev->needed_tailroom;
+	needed_tailroom -= skb_tailroom(skb);
+	needed_tailroom = max_t(int, needed_tailroom, 0);
+
+	if (pskb_expand_head(skb, r92su->wdev.netdev->needed_headroom,
+			     needed_tailroom, GFP_ATOMIC)) {
+		return TX_DROP;
+	}
 
 	/* clean up tx info */
 	memset(tx_info, 0, sizeof(tx_info));
